mempool/cat/pool.go

```diff
@@ -0,0 +1,733 @@
+		return rsp, err
+	}
+	if rsp.Code != abci.CodeTypeOK {
+		txmp.rejectedTxCache.Push(key)
+		txmp.metrics.FailedTxs.Add(1)
+		return rsp, fmt.Errorf("application rejected transaction with code %d", rsp.Code)
+	}
+
+	// Create wrapped tx
+	wtx := newWrappedTx(
+		tx, key, txmp.height, rsp.GasWanted, rsp.Priority, rsp.Sender,
+	)
+
+	// Perform the post check
+	err = txmp.postCheck(wtx.tx, rsp)
+	if err != nil {
+		txmp.rejectedTxCache.Push(key)
+		txmp.metrics.FailedTxs.Add(1)
+		return rsp, fmt.Errorf("rejected bad transaction after post check: %w", err)
+	}
+
+	// Now we consider the transaction to be valid. Once a transaction is valid, it
+	// can only become invalid if recheckTx is enabled and RecheckTx returns a non zero code
+	if err := txmp.addNewTransaction(wtx, rsp); err != nil {
+		return nil, err
+	}
+	return rsp, nil
+}
+
+// RemoveTxByKey removes the transaction with the specified key from the
+// mempool. It reports an error if no such transaction exists. This operation
+// does not remove the transaction from the rejectedTxCache.
+func (txmp *TxPool) RemoveTxByKey(txKey types.TxKey) error {
+	txmp.rejectedTxCache.Push(txKey)
+	_ = txmp.evictedTxs.Pop(txKey)
+	_ = txmp.store.remove(txKey)
+	txmp.seenByPeersSet.RemoveKey(txKey)
+	txmp.metrics.EvictedTxs.Add(1)
+	return nil
+}
+
+// Flush purges the contents of the mempool and the cache, leaving both empty.
+// The current height is not modified by this operation.
+func (txmp *TxPool) Flush() {
+	// Remove all the transactions in the list explicitly, so that the sizes
+	// and indexes get updated properly.
+	size := txmp.Size()
+	txmp.store.reset()
+	txmp.seenByPeersSet.Reset()
+	txmp.evictedTxs.Reset()
+	txmp.rejectedTxCache.Reset()
+	txmp.metrics.EvictedTxs.Add(float64(size))
+}
+
+// PeerHasTx marks that the transaction has been seen by a peer.
+// It returns true if the mempool has the transaction and has recorded the
+// peer and false if the mempool has not yet seen the transaction that the
+// peer has
+func (txmp *TxPool) PeerHasTx(peer uint16, txKey types.TxKey) {
+	txmp.logger.Debug("peer has tx", "peer", peer, "txKey", fmt.Sprintf("%X", txKey))
+	txmp.seenByPeersSet.Add(txKey, peer)
+}
+
+// allEntriesSorted returns a slice of all the transactions currently in the
+// mempool, sorted in nonincreasing order by priority with ties broken by
+// increasing order of arrival time.
+func (txmp *TxPool) allEntriesSorted() []*wrappedTx {
+	txs := txmp.store.getAllTxs()
+	sort.Slice(txs, func(i, j int) bool {
+		if txs[i].priority == txs[j].priority {
+			return txs[i].timestamp.Before(txs[j].timestamp)
+		}
+		return txs[i].priority > txs[j].priority // N.B. higher priorities first
+	})
+	return txs
+}
+
+// ReapMaxBytesMaxGas returns a slice of valid transactions that fit within the
+// size and gas constraints. The results are ordered by nonincreasing priority,
+// with ties broken by increasing order of arrival.  Reaping transactions does
+// not remove them from the mempool.add
```

Comment:
Is the `.add` intentional? If not:

```suggestion
// not remove them from the mempool
```